webui/react/src/pages/F_ExpList/F_ExperimentList.tsx

```diff
@@ -0,0 +1,185 @@
+
+import GlideTable from './glide-table/GlideTable';
+import { useGlasbey } from './glide-table/useGlasbey';
+
+const experimentColumns = [
+  'archived',
+  'checkpointCount',
+  'checkpointSize',
+  'description',
+  'duration',
+  'forkedFrom',
+  'id',
+  'name',
+  'progress',
+  'resourcePool',
+  'searcherType',
+  'searcherMetricValue',
+  'selected',
+  'startTime',
+  'state',
+  'tags',
+  'numTrials',
+  'user',
+] as const;
+
+export type ExperimentColumn = (typeof experimentColumns)[number];
+
+const defaultExperimentColumns: ExperimentColumn[] = [
+  'id',
+  'description',
+  'tags',
+  'forkedFrom',
+  'progress',
+  'startTime',
+  'state',
+  'searcherType',
+  'user',
+  'duration',
+  'numTrials',
+  'resourcePool',
+  'checkpointSize',
+  'checkpointCount',
+  'searcherMetricValue',
+];
+
+interface Props {
+  project: Project;
+}
+
+const emptyExperiment: Omit<ExperimentItem, 'config' | 'configRaw'> = {
+  archived: false,
+  checkpointCount: 0,
+  checkpointSize: 0,
+  description: '',
+  endTime: '2021-06-21T18:11:23.756443024Z',
+  forkedFrom: undefined,
+  hyperparameters: {},
+  id: 1,
+  jobId: 'backfilled-1',
+  labels: [],
+  name: '',
+  notes: '',
+  numTrials: 0,
+  projectId: 0,
+  resourcePool: '',
+  searcherType: 'single',
+  startTime: '2020-06-26T18:09:25.844705105Z',
+  state: 'COMPLETED',
+  trialIds: [],
+  userId: 1,
+};
+
+const PAGE_RADIUS = 50;
+const F_ExperimentList: React.FC<Props> = ({ project }) => {
+  const [pageMidpoint, setPageMidpoint] = useState(0);
+  const [experiments, setExperiments] = useState<ExperimentItem[]>([]);
+  const [sortableColumnIds, setSortableColumnIds] = useState(defaultExperimentColumns);
+  const [selectedExperimentIds, setSelectedExperimentIds] = useState<string[]>([]);
+  const [selectAll, setSelectAll] = useState(false);
+  const [isLoading, setIsLoading] = useState(true);
+  const [canceler] = useState(new AbortController());
+
+  const colorMap = useGlasbey(selectedExperimentIds);
+  const pageRef = useRef<HTMLElement>(null);
+  const { width } = useResize(pageRef);
+
+  const handleScroll = useCallback(({ y, height }: Rectangle) => {
+    const visibleRegionMidPoint = y + height / 2;
+    setPageMidpoint(Math.round(visibleRegionMidPoint / PAGE_RADIUS) * PAGE_RADIUS);
+  }, []);
+
+  const fetchExperiments = useCallback(async (): Promise<void> => {
+    try {
+      const tableLimit = 2 * PAGE_RADIUS;
+      const tableOffset = Math.max(pageMidpoint - PAGE_RADIUS, 0);
+
+      const response = await getExperiments(
+        {
+          archived: false,
+          limit: tableLimit,
+          offset: tableOffset,
+          projectId: project.id,
+        },
+        { signal: canceler.signal },
+      );
+
+      setExperiments((prevExperiments) => {
+        const paddedExperimentBeforeCurrentPage = [
+          ...prevExperiments.slice(0, tableOffset),
+          ...Array(Math.max(0, tableOffset - prevExperiments.length)).fill(emptyExperiment),
```

Comment:
Shouldn't this just be filled with `undefined`? That was my understanding of what the `while` code was doing.

Reply:
that's what it was doing, but there was a comment calling that we should coalesce to an empty experiment.

basically to prevent a possible page crash.

Reply:
it's enough of an edge case and local to this part of the code that I prefer to do it this way rather than muddying the waters elsewhere

Reply:
It's only local to this until the invalid assumptions cause an error elsewhere.

Reply:
like, i don't think this ever _should_ happen, but i am not 100%, and would rather just do a small piece of defensive coding than to propagate a bunch of optionals

Reply:
> It's only local to this until the invalid assumptions cause an error elsewhere.

well the worst thing that happens in this case is that you scroll back up and see some empty rows while the data is being refreshed

Reply:
If it really will not ever happen then we shouldn't handle this case. Handling cases that _shouldn't_ happen just masks bugs and obfuscates the code. You could leave a comment saying what invariant you expect to be true here or even throw an exception if that invariant is violated to make sure the bug is noticed.

Reply:
> well the worst thing that happens in this case is that you scroll back up really fast and see some empty rows while the data is being refreshed

If that's something we need to handle we should handle that loading state _explicitly_ in the row renderer.